lib/spack/spack/spec.py

```diff
@@ -2799,13 +2800,40 @@ def format(self, format_string='$_$@$%@+$+$=', **kwargs):
 
         Anything else is copied verbatim into the output stream.
 
-        *Example:*  ``$_$@$+`` translates to the name, version, and options
-        of the package, but no dependencies, arch, or compiler.
+        Args:
+            format_string (str): string containing the format to be expanded
+
+            **kwargs (dict): the following list of keywords is supported
+
+                - color (bool): True if returned string is colored
+
+                - transform (dict): maps full-string formats to a callable \
+                that accepts a string and returns another one
+
+        Examples:
+
+            The following line:
+
+            .. code-block:: python
+
+                s = spec.format('$_$@$+')
+
+            translates to the name, version, and options of the package, but no
+            dependencies, arch, or compiler.
 
         TODO: allow, e.g., ``$6#`` to customize short hash length
         TODO: allow, e.g., ``$//`` for full hash.
         """
         color = kwargs.get('color', False)
+
+        # Set up a dictionary of transformations for named tokens.
+        # The default is to leave the string unchanged.
+        #
+        # The expression `lambda: lambda x: x` is the shortest and neatest
+        # way to have a factory of identity functions in python
+        tr = collections.defaultdict(lambda: lambda x: x)
```

Comment:
(this applies if you disagree that a whitelist would be more suitable)

I'd prefer this had a more descriptive name like `token_transforms` and right after `named_str` is defined below do something like:

`token_transform = token_transforms.get(named_str)`

Also I think doing the following would avoid the defaultdict with `lambda: lambda x: x`:

`token_transform = token_transforms.get(named_str, lambda x: x)`

Reply:
It won't be a problem to change the naming, if if you strongly disagree with it. However: here the scope of the function is relatively contained, and making this name longer will give a lot more line breaks due to `pep8`. I tried having a longer name along the line of what you suggest, but in the end I went back to this short one because I was not convinced there was any readability improvement.

> Also I think doing the following would avoid the defaultdict

Yes. That would avoid `defaultdict`. Not that it is a big deal, but why do you think this is preferable?

Reply:
Done - to be honest, despite I see no harm in this transformation (`tr` -> `token_transforms`, `defaultdict` - > `dict.get(value, default)`) I also see no gain.

> generally I think line breaks can be avoided by adding lines (e.g. defining variables)

Sure. Another possibility is to use shorter names for variables whose lifetime is also short - if what they mean can be inferred by the context.

> (for example you included an explanatory comment of what it was doing).

I added the comment because I was using lambdas, and generally people seem to be wary of them. I think if somebody doesn't need a comment to get that:
```python
lambda x: x
```
is the identity function, then he won't need a comment to get that:
```python
lambda: lambda x: x
```
is a factory of identity functions. Just my opinion though.